### PR TITLE
COMPAT: Change khcomplex64/128_t typedefs

### PR DESCRIPTION
**Pull request by mroeschke**
xref https://github.com/numpy/numpy/pull/24085

suggestion by @seberg

### CODE REVIEW

**Review by lithomas1 — APPROVED** [review]

**Review by WillAyd — APPROVED** [review]

File: pandas/_libs/include/pandas/vendored/klib/khash_python.h [hunk 1]
```diff
@@ -2,10 +2,14 @@
...
 
 
-// use numpy's definitions for complex
-#include <numpy/arrayobject.h>
-typedef npy_complex64 khcomplex64_t;
-typedef npy_complex128 khcomplex128_t;
+typedef struct {
+    float real;
+    float imag;
+    } khcomplex64_t;
```
Comment (WillAyd) [inline pandas/_libs/include/pandas/vendored/klib/khash_python.h:8]:
```suggestion
} khcomplex64_t;
```

Just a minor readability thing if you have to edit again. We should probably run this whole file through `clang-format` in a separate PR / CI

File: pandas/_libs/include/pandas/vendored/klib/khash_python.h [hunk 2]
```diff
@@ -2,10 +2,14 @@
...
+typedef struct {
+    double real;
+    double imag;
```
Comment (jbrockmendel) [inline pandas/_libs/include/pandas/vendored/klib/khash_python.h:11]: will this change anything on platforms where float/double aren't 32/64 bits?  (which i guess we have zero testing for so not a huge deal)

Comment (mroeschke) [reply pandas/_libs/include/pandas/vendored/klib/khash_python.h:11]: Not sure. It appears in https://github.com/numpy/numpy/pull/24085 in numpy/core/include/numpy/npy_math.h numpy also defines real/imag components as doubles

Comment (WillAyd) [reply pandas/_libs/include/pandas/vendored/klib/khash_python.h:11]: I think the misnomer is to call these complex64 and complex128. The C99 standard has float complex and double complex types, but nothing that is a fixed width

See also https://stackoverflow.com/questions/2524737/fixed-size-floating-point-types#2524933

**Comment by mroeschke** [comment]
Can't reproduce this locally, but it seems that `float real` of `complex(1.2, nan)` is returning NaN instead of 1.2. Any ideas @WillAyd

File: pandas/_libs/include/pandas/vendored/klib/khash_python.h (hunk 2 above)
Comment (seberg) [reply pandas/_libs/include/pandas/vendored/klib/khash_python.h:11]: Yeah, I guess technically float32/float64 are sized aliases under the assumption they refer to float/double, but in practice the assumption seems so strong that I doubt it matters.

I am very puzzled about the change though, this is exactly what it was after all, and I don't think C cares about type aliases (even if there was interaction anywhere).  I am wondering if the `include <numpy/arrayobject.h>` somehow is involved, but I don't see _why_ it should be either?

I assume the value stored is correct here and just the printing somehow thinks it is printing a NaN?

**Comment by lithomas1** [comment]
I think it might be a Python 3.12 issue. The regular numpy dev build doesn't fail.

**Comment by mroeschke** [comment]
> I think it might be a Python 3.12 issue. The regular numpy dev build doesn't fail.

I think numpy dev has cached dependencies so maybe the changes haven't caught up

**Comment by mroeschke** [comment]
Looks like this change subtly changed how `maybe_convert_objects` infers complex scalars, but given that the failing test is a net new fix in 2.1 and the failure is an edge case with nan, I think the status quo is better that the behavior in 2.0 so merging as is.

**Pull request merged**